cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -641,29 +645,76 @@ traceBlockFetchClientMetrics
   :: forall blk remotePeer.
      ( )
   => Maybe EKGDirect
-  -> STM.TVar SlotNo
+  -> STM.TVar (SMap.Map SlotNo NominalDiffTime)
+  -> STM.TVar Int64
+  -> STM.TVar Int64
+  -> STM.TVar Int64
   -> Tracer IO (TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)))
   -> Tracer IO (TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)))
-traceBlockFetchClientMetrics Nothing _ tracer = tracer
-traceBlockFetchClientMetrics (Just ekgDirect) lastSlotNoVar tracer = Tracer $ bfTracer
+traceBlockFetchClientMetrics Nothing _ _ _ _ tracer = tracer
+traceBlockFetchClientMetrics (Just ekgDirect) slotMapVar cdf1sVar cdf2sVar cdf5sVar tracer = Tracer bfTracer
 
   where
+
+    -- We track the cdf for 1s, 2s and 5s.
+    updateCDFCounter :: NominalDiffTime -> Int -> Int64 -> STM (Double, Double, Double)
+    updateCDFCounter delay size step = do
+
+      when (delay < 1) $ do
+        STM.modifyTVar' cdf1sVar (+ step)
+      when (delay < 2) $ do
+        STM.modifyTVar' cdf2sVar (+ step)
+      when (delay < 5) $ do
+        STM.modifyTVar' cdf5sVar (+ step)
+
+      cdf1s <- STM.readTVar cdf1sVar
+      cdf2s <- STM.readTVar cdf2sVar
+      cdf5s <- STM.readTVar cdf5sVar
+      return (fromIntegral cdf1s / fromIntegral size, fromIntegral cdf2s / fromIntegral size,
+              fromIntegral cdf5s / fromIntegral size)
+
+
     bfTracer :: TraceLabelPeer remotePeer (TraceFetchClientState (Header blk)) -> IO ()
     bfTracer e@(TraceLabelPeer _ (CompletedBlockFetch p _ _ _ delay)) = do
       traceWith tracer e
-      let slotNo = case pointSlot p of
-                        Origin -> SlotNo 0 -- We don't care
-                        At s   -> s
-      fresh <- atomically $ do
-          lastSlotNo <- STM.readTVar lastSlotNoVar
-          if lastSlotNo < slotNo
-             then do
-                 STM.writeTVar lastSlotNoVar slotNo
-                 return True
-             else return False
-      when fresh $
-        sendEKGDirectDouble ekgDirect "cardano.node.metrics.blockfetchclient.blockdelay.s"
-            $ realToFrac delay
+      case pointSlot p of
+        Origin -> return () -- Nothing to do.
+        At slotNo -> do
+          (fresh, cdf1s, cdf2s, cdf5s) <- atomically $ do
+              slotMap <- STM.readTVar slotMapVar
+              case SMap.lookup slotNo slotMap of
+                   Nothing -> do
+                     let slotMap' = SMap.insert slotNo delay slotMap
```

Comment:
Note that `SMap` here is the non-strict `Data.Map`. We probably want the `Data.Map.Strict`

Reply:
I've switched the SMap import to use Data.Map.Strict.